src/testing/creation.rs

```diff
@@ -0,0 +1,293 @@
+#[tokio::test(flavor = "multi_thread", worker_threads = 3)]
+async fn disconnected_creators_should_create_dag() {
+    let n_members = NodeCount(7);
+    let max_round: Round = 25;
+
+    let (mut test_controller, killers, handles, mut units_from_controller, units_for_creators) =
+        setup_test(n_members);
+    loop {
+        futures::select! {
+            _ = test_controller.control_until(max_round).fuse() => break,
+            unit = units_from_controller.next() => match unit {
+                Some(unit) => for units_for_creator in &units_for_creators {
+                    units_for_creator.unbounded_send(unit.clone()).expect("Channel to creator should be open");
+                },
+                None => panic!("Channel from controller should be open."),
+            }
+        }
+    }
+    let max_round: Round = 50;
+    let mut dropped_units = Vec::new();
+    loop {
+        futures::select! {
+            _ = test_controller.control_until(max_round).fuse() => break,
+            unit = units_from_controller.next() => match unit {
+                Some(unit) => {
+                    for units_for_creator in units_for_creators.iter().skip(2) {
+                        units_for_creator.unbounded_send(unit.clone()).expect("Channel to creator should be open");
+                    }
+                    dropped_units.push(unit);
+                },
+                None => panic!("Channel from controller should be open."),
+            }
+        }
+    }
+    let max_round: Round = 75;
+    for unit in dropped_units {
+        for units_for_creator in units_for_creators.iter().take(2) {
+            units_for_creator
+                .unbounded_send(unit.clone())
+                .expect("Channel to creator should be open");
+        }
+    }
+    loop {
+        futures::select! {
+            _ = test_controller.control_until(max_round).fuse() => break,
+            unit = units_from_controller.next() => match unit {
+                Some(unit) => for units_for_creator in &units_for_creators {
+                    units_for_creator.unbounded_send(unit.clone()).expect("Channel to creator should be open");
+                },
+                None => panic!("Channel from controller should be open."),
+            }
+        }
+    }
+    assert!(test_controller
+        .max_round_per_creator
+        .iter()
+        .all(|r| *r >= (max_round - 3)));
+    assert!(
+        test_controller
+            .max_round_per_creator
+            .iter()
+            .filter(|r| *r >= &(max_round - 1))
+            .count()
+            >= 5
+    );
+    finish(killers, handles).await;
+}
+
+// Catching up test
+// This test checks if 5 creators that start at the same time and 2 creators
+// that start after those first 5 reach round 25,
+// will reach at least round 47, when at least one reaches round 50.
+#[tokio::test(flavor = "multi_thread", worker_threads = 3)]
+async fn late_creators_should_create_dag() {
```

Comment:
The idea is ok but its execution is a bit tricky. We should avoid such tests. I am not sure, but it looks like it can fail non-deterministically. I think we have (almost?) all ingredients in place to write these tests `deterministically`.

Reply:
I don't think we can test for this deterministically – after all we are essentially checking only whether specific delays are skipped. I agree that this test is not ideal (and technically can fail), but I don't see how else we can check this, especially with `is_behind` now removed.

Reply:
Since proposed solution is far from perfect and a good one requires more refactoring, we decided to leave it for future improvement - A0-453 in our jira.